pkg/image/registryclient/alternativesources.go

```diff
@@ -0,0 +1,170 @@
+	if err != nil {
+		return fmt.Errorf("error decoding ImageContentSourcePolicy from %s: %v", c.ImageSources.icspFile, err)
+	}
+	icsp, ok := icspObj.(*operatorv1alpha1.ImageContentSourcePolicy)
+	if !ok {
+		return fmt.Errorf("could not decode ImageContentSourcePolicy from %s", c.ImageSources.icspFile)
+	}
+	c.ImageSources.icspList = append(c.ImageSources.icspList, *icsp)
+	return nil
+}
+
+// alternativeImageSources returns list of DockerImageReference objects from list of ImageContentSourcePolicy objects
+func (a *Context) alternativeImageSources(imageRef imagereference.DockerImageReference) ([]imagereference.DockerImageReference, error) {
+	var imageSources []imagereference.DockerImageReference
+	for _, icsp := range a.ImageSources.icspList {
+		repoDigestMirrors := icsp.Spec.RepositoryDigestMirrors
+		var sourceMatches bool
+		for _, rdm := range repoDigestMirrors {
+			rdmRef, err := imagereference.Parse(rdm.Source)
+			if err != nil {
+				return nil, err
+			}
+			if imageRef.AsRepository() == rdmRef.AsRepository() {
+				klog.V(2).Infof("%v RepositoryDigestMirrors source matches given image", imageRef.AsRepository())
+				sourceMatches = true
+			}
+			for _, m := range rdm.Mirrors {
+				if sourceMatches {
+					klog.V(2).Infof("%v RepositoryDigestMirrors mirror added to potential ImageSourcePrefixes from ImageContentSourcePolicy", m)
+					mRef, err := imagereference.Parse(m)
+					if err != nil {
+						return nil, err
+					}
+					imageSources = append(imageSources, mRef)
+				}
+			}
+		}
+	}
+	uniqueMirrors := make([]imagereference.DockerImageReference, 0, len(imageSources))
+	uniqueMap := make(map[imagereference.DockerImageReference]bool)
+	for _, imageSourceMirror := range imageSources {
+		if _, ok := uniqueMap[imageSourceMirror]; !ok {
+			uniqueMap[imageSourceMirror] = true
+			uniqueMirrors = append(uniqueMirrors, imageSourceMirror)
+		}
+	}
+	// make sure at least 1 imagesource
+	// ie, make sure the image passed is included in image sources
+	// this is so the user-given image ref will be tried
+	if len(imageSources) == 0 {
+		imageSources = append(imageSources, imageRef.AsRepository())
+		return imageSources, nil
+	}
+	klog.V(2).Infof("Found sources: %v for image: %v", uniqueMirrors, imageRef)
+	return uniqueMirrors, nil
+}
+
+// PreferredImageSource chooses appropriate DockerImageReference for a Context from possible image sources gathered from
+// ImageContentSourcePolicy objects and user-passed image. PreferredImageSource will lookup from cluster
+// if alwaysTryAlternativeSources is true, otherwise will lookup from file if file given, or from
+// original image-reference of user-given image (that may be different from user-given in case of mirrored images).
+func (c *Context) PreferredImageSource(image imagereference.DockerImageReference, alwaysTryAlternativeSources, insecure bool) (imagereference.DockerImageReference, distribution.Repository, distribution.ManifestService, error) {
```

Comment:
As commented in the other PR, this isn't the interface we want (defeats the purpose of putting it here).  We put it here to hide the details of alternates behind the interface.

/hold

I think we should probably set aside time and walk through a basic impl that would accomplish that goal - I may not be clearly expressing the limitations I think should be in place for this, but we want to encapsulate and hide ICSP, not expose it.

Reply:
OK, lemme try to work this out with @sallyom and we'll ping you again. I might have misunderstood or forgotten between when we last spoke and now.

Reply:
i pushed an update ptal thanks